test/common/config/grpc_mux_impl_test.cc

```diff
@@ -722,6 +723,190 @@ TEST_F(GrpcMuxImplTest, BadLocalInfoEmptyNodeName) {
       "ads: node 'id' and 'cluster' are required. Set it either in 'node' config or via "
       "--service-node and --service-cluster options.");
 }
+
+// Send discovery request with v2 resource type_url, receive discovery response with v3 resource
+// type_url.
+TEST_F(GrpcMuxImplTest, WatchV2ResourceV3) {
+  TestScopedRuntime scoped_runtime;
+  Runtime::LoaderSingleton::getExisting()->mergeValues(
+      {{"envoy.reloadable_features.enable_type_url_downgrade_and_upgrade", "true"}});
+  setup();
+  InSequence s;
+  TestUtility::TestOpaqueResourceDecoderImpl<envoy::config::endpoint::v3::ClusterLoadAssignment>
+      resource_decoder("cluster_name");
+  const std::string& v2_type_url = Config::TypeUrl::get().ClusterLoadAssignment;
+  const std::string& v3_type_url =
+      "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment";
+  NiceMock<MockSubscriptionCallbacks> foo_callbacks;
+  auto foo_sub = grpc_mux_->addWatch(v2_type_url, {"x", "y"}, foo_callbacks, resource_decoder);
+  NiceMock<MockSubscriptionCallbacks> bar_callbacks;
+  auto bar_sub = grpc_mux_->addWatch(v2_type_url, {"y", "z"}, bar_callbacks, resource_decoder);
+  EXPECT_CALL(*async_client_, startRaw(_, _, _, _)).WillOnce(Return(&async_stream_));
+  expectSendMessage(v2_type_url, {"y", "z", "x"}, "", true);
+  grpc_mux_->start();
+
+  {
+    // Send resource with v3 type url, should invoke onConfigUpdate.
+    auto response = std::make_unique<envoy::service::discovery::v3::DiscoveryResponse>();
+    response->set_type_url(v3_type_url);
+    response->set_version_info("1");
+    envoy::config::endpoint::v3::ClusterLoadAssignment load_assignment;
+    load_assignment.set_cluster_name("x");
+    response->add_resources()->PackFrom(load_assignment);
+    EXPECT_CALL(bar_callbacks, onConfigUpdate(_, "1")).Times(0);
+    EXPECT_CALL(foo_callbacks, onConfigUpdate(_, "1"))
+        .WillOnce(Invoke([&load_assignment](const std::vector<DecodedResourceRef>& resources,
+                                            const std::string&) {
+          EXPECT_EQ(1, resources.size());
+          const auto& expected_assignment =
+              dynamic_cast<const envoy::config::endpoint::v3::ClusterLoadAssignment&>(
+                  resources[0].get().resource());
+          EXPECT_TRUE(TestUtility::protoEqual(expected_assignment, load_assignment));
+        }));
+    // Send request with v2 type url resources.
+    expectSendMessage(v2_type_url, {"y", "z", "x"}, "1");
+    grpc_mux_->grpcStreamForTest().onReceiveMessage(std::move(response));
+  }
+
+  {
```

Comment:
What role does this second block play in testing (same for test below)? All things being equal, shorter tests for same line/functional coverage are better.

Reply:
Thanks, I have removed the second block.